design-documents/service-isolation/caching-layer.md

```diff
@@ -0,0 +1,105 @@
+
+[BFF](https://github.com/magento/architecture/blob/master/design-documents/service-isolation.md#backends-for-frontends) http-responses contains static assets, **public** and **private** dynamic content.
+Static assets should be cached using CDN. Reverse proxy should cache public content.
+It could be Varnish (acts as reverse proxy) or Fastly (combines reverse proxy and CDN functions).
+Varnish and Fastly are already integrated with Magento. Reverse proxy uses lazy caching approach.
+Beside performance this type of cache has next benefits:
+ * Protection against outages - can optionally serve stale content when there is a problem with origin server
+ * Scalability - number of caching nodes can be increased
+ * Flexibility – Varnish Configuration Language (VCL) builds customized solutions, rules and modules
+ * Shielding - the shield node is the only node to get the content from origin server, other cache nodes will fetch from the shield node
+ 
+ ![Public content caching](img/public-cache.jpg)
+
+A private or visitor-specific content should be stored on client-side (browser). After POST requests the cache 
+in the web browser will be flushed and an AJAX call will be done to fetch an updated copy of the private content.
+Also private content cache on client-side should expire according to TTL.
+
+ ![Private content caching](img/private-cache.jpg)
+ 
+ In general, the same caching approach works in current monolith Magento architecture.
+ For service oriented architecture next issues should be solved:
+  * The http-response header should include a set of tags for effective cache purging on reverse-proxy.
+ The set should be created while request is going via the chain of services. Appropriate design should be prepared.
+  * Caching on reverse-proxy should be aligned with request authorization approach. The presence of an authorization token should be taken into account.
+  
+
+### Application data caching
+
+Service application needs to cache such data as DB query results, merged configurations for efficiently and rapidly respond to requests.
+
+The three possible types of caches are the following:
+ * Local caches
+ * Remote caches dedicated to each service
+ * Remote caches shared across the services
+ 
+#### Local caches
+ - Pros: fast, no network traffic associated with retrieving data
+ - Cons: 
+   - shares sources with application
+   - can lead to data inconsistency when multiple instances of application servers are created
+   - the information stored within an individual cache node cannot be shared with other application servers
+   - not follow the 12-factor principles (any data that needs to persist must be stored in a stateful backing service)
+ 
+ ![Local cache](img/local-app-cache.png)
+ 
+#### Remote caches dedicated to each service
+ - Pros: 
+   - scalable, adheres to 12-factor principles
+   - externalize application state
+   - possibility dynamically change the number of application instances
```

Comment:
"application instances" - I assume you mean "service instances"?

Reply:
Yes, it will be more accurate. Fixed, thanks